### ISSUE TITLE
Yield error when operator section is used in statement context

### ISSUE THREAD

**Issue by jdunkerley** [enso-org/enso#11203]
```ruby
from Standard.Base import all

main =
    test = [1] + [2] + [3] + [4]
    test.to_text
```
results in [1,2,3,4]

however

```ruby
from Standard.Base import all

main =
    test = [1] +
        [2] +
        [3] +
        [4]
    test.to_text
```
results in `[1, 4]`

**Update** by @JaroslavTulach based on [Kaz's comment](https://github.com/enso-org/enso/issues/11203#issuecomment-2386480463): I never realized that the _prefix syntax_ works for multiline expressions without any issues:
```ruby
from Standard.Base import all

main = [1]
    + [2]
    + [3]
    + [4]
```
prints `[1, 2, 3, 4]`.

**Comment by kazcw** [comment]
- This is the same syntactic pitfall as here: #9944

We could improve this by emitting a warning when an expression with no side effects is evaluated without being used. That might be a tall order for the backend to implement in general, but this is not the first time a bug has been reported due to this pitfall--to catch this specific case, we could have a warning if an operator-section occurs in statement context. This would be a syntax warning, so the parser would be the place to implement it.

**Comment by kazcw** [comment]
Found the previous issue: #9944

**Comment by JaroslavTulach** [comment]
- a related issue is also tracked as #8024

> improve this by emitting a warning when an expression with no side effects is evaluated without being used. 

- such a warning has been requested at #5430 - I can emit it, but I have no chance to find out if or not the expression had some side effects - e.g. it may be a false alarm.

**Comment by JaroslavTulach** [comment]
> not the first time a bug has been reported due to this pitfall

@kazcw, why don't we change the [Body code blocks specification](https://github.com/enso-org/design/blob/wip/wd/enso-spec/epics/enso-spec-1.0/03.%20Code%20format%20and%20layout.md#body-code-blocks) to say that a block where lines end with an operator like `+`:
```ruby
from Standard.Base import all

main =
    test = [1] +
        [2] +
        [3] +
        [4]
    test.to_text
```
is equivalent to concatenation - e.g. `[1] + [2] + [3] + [4]`?

**Comment by kazcw** [comment]
That's a good question. A rule change could allow recognizing this case:
- Only `=` or `->` introduces a body block (currently any operator can take a body block as an operand, but this is not used in practice).
- Any other operator may introduce a *postfix operator block*.

---

There is a clear argument for this: Two **language users have already expected it to work**; maybe it should.

However my opinion, from the perspective of the syntax rules, is that supporting this would **make writing easier at the expense of greater complexity in reading and learning the language**.

- **The difference from prefix operator blocks looks like a style difference, but has a subtle semantic difference in associativity**: Current (*prefix*) operator blocks are evaluated with top-to-bottom line associativity; the natural associativity of a postfix operator block would be bottom-to-top (or potentially as if lines were concatenated?). There may be cases where having a choice of associativity would be useful, but I think more often it would be a stumbling block.

- **The *postfix* syntax doesn't fit into the language as well** as the current syntax. This is more subjective, but I think the postfix syntax could impede an intuitive sense of how blocks and operator sections work. In the prefix syntax, all the indented lines have the same shape, making it more obvious that it is a special block type; but in the postfix syntax the first operator belongs to the parent line, and subsequent operators belong to all the non-final lines--even though the first operator and later operators are logically "equals" in their roles in the AST.

---

But, more important than a language implementer's opinion is the perspective of the people writing Enso every day, so I have a question for @jdunkerley (CC @radeusgd):

Would being able to write expressions this way be useful enough to justify a bit more language complexity? Or would it be better to have only one way to do it, and a parser warning directing users toward the prefix syntax?

**Comment by JaroslavTulach** [comment]
OK, let's **stick with prefix syntax** for now and implement runtime detection of these discarded values
- #5430

However @radeusgd would prefer a static check. @kazcw, maybe the _postfix syntax_ doesn't make much sense from the language perspective to be implemented, but would it **make a sense to detect it**? Report it as a _warning_? What does it mean when line ends with an operator anyway? I tried:
```ruby
> a f = f 2 5
> a (*)
>>> 10
```
but that's an operator surrounded by parenthesis! However at the end I found valid code that ends the line with `*`:
```ruby
> a f = f 5
> a <| 2 *
10
```
possibly the `a` function could have some side-effects, so it seems to me (again) there is no way to statically detect whether the line ending with an operator has a meaning or not.

**Comment by radeusgd** [comment]
> However at the end I found valid code that ends the line with `*`:
> 
> ```ruby
> > a f = f 5
> > a <| 2 *
> 10
> ```
> 
> possibly the `a` function could have some side-effects, so it seems to me (again) there is no way to statically detect whether the line ending with an operator has a meaning or not.

But in this valid code, the result of `2 *` is **not discarded**.

What if we decide to create a convention that operators like `*` should not have side-effects, and thus we could warn when their usage is discarded? Of course this would only be a convention, but then it would only be a warning and not an error.

**Comment by kazcw** [comment]
> so it seems to me (again) there is no way to statically detect whether the line ending with an operator has a meaning or not.

Yes, it's impossible to detect all cases statically. But a syntax warning for one case would go a long way: An operator section in statement context is always dead code. I suspect detecting anything that parses as such would have caught the issue here and in #9944 as soon as it occurred--it certainly would catch the issue in the simplified code included in the bug reports.

**Comment by radeusgd** [comment]
> > so it seems to me (again) there is no way to statically detect whether the line ending with an operator has a meaning or not.
> 
> Yes, it's impossible to detect all cases statically. But a syntax warning for one case would go a long way: An operator section in statement context is always dead code. I suspect detecting anything that parses as such would have caught the issue here and in #9944 as soon as it occurred--it certainly would catch the issue in the simplified code included in the bug reports.

That's exactly what I wanted to suggest.

**Comment by JaroslavTulach** [comment]
> An operator section in statement context is always dead code.

- @kazcw, can I detect that in the `TreeToIr`? How?
- or does that need to be detected on the Rust side of the parser?

**Comment by kazcw** [comment]
@JaroslavTulach
> > An operator section in statement context is always dead code.
> 
>     * [@kazcw](https://github.com/kazcw), can I detect that in the `TreeToIr`? How?
> 
>     * or does that need to be detected on the Rust side of the parser?

It's a change in the subset of nodes that may be the child of an Expression Statement--it sounds like a syntax rule to me, and treating it that way should simplify some logic in the backend and frontend, once the Expression Statement has typesafe bindings (part of #10754). The change is simple, mostly a matter of testing etc. I could do it tomorrow.

**Comment by JaroslavTulach** [comment]
- It would be great to finally close this endeavour
- there was a lot of issues related to this _operator/block mismatch_
- driving this one to some conclusion would sufficiently close the reports, I hope
